contracts/contracts/GrantRoundManager.sol

```diff
@@ -99,66 +126,119 @@ contract GrantRoundManager {
   }
 
   /**
-   * @notice Swap and donate to a grant
-   * @param  _donation Donation being made to a grant
+   * @notice Performs swaps if necessary and donates funds as specified
+   * @param _swaps Array of SwapSummary objects describing the swaps required
+   * @param _deadline Unix timestamp after which a swap will revert, i.e. swap must be executed before this
+   * @param _donations Array of donations to execute
+   * @dev `_deadline` is not part of the `_swaps` array since all swaps can use the same `_deadline` to save some gas
+   * @dev Caller must ensure the input tokens to the _swaps array are unique
+   * @dev Does not verify
    */
-  function swapAndDonate(Donation calldata _donation) external payable {
-    // --- Validation ---
-    // Rounds must be specified
-    GrantRound[] calldata _rounds = _donation.rounds;
-    require(_rounds.length > 0, "GrantRoundManager: Must specify at least one round");
-
-    // Only allow value to be sent if the input token is WETH (this limitation should be fixed in #76, as this
-    // require statement prohibits donating WETH)
-    IERC20 _tokenIn = IERC20(_donation.path.toAddress(0));
-    require(
-      (msg.value == 0 && address(_tokenIn) != WETH) || (msg.value > 0 && address(_tokenIn) == WETH),
-      "GrantRoundManager: Invalid token-value pairing"
-    );
+  function donate(
+    SwapSummary[] calldata _swaps,
+    uint256 _deadline,
+    Donation[] calldata _donations
+  ) external payable {
+    // Main logic
+    _validateDonations(_donations);
+    _executeDonationSwaps(_swaps, _deadline);
+    _transferDonations(_donations);
+
+    // Clear storage for refunds (this is set in _executeDonationSwaps)
+    for (uint256 i = 0; i < _swaps.length; i++) {
+      IERC20 _tokenIn = IERC20(_swaps[i].path.toAddress(0));
+      swapOutputs[_tokenIn] = 0;
+      donationRatios[_tokenIn] = 0;
+    }
+  }
 
-    // Ensure grant recieving donation exists in registry
-    uint96 _grantId = _donation.grantId;
-    require(_grantId < registry.grantCount(), "GrantRoundManager: Grant does not exist in registry");
-
-    // Iterate through each GrantRound to verify:
-    //   - The round has the same donationToken as the GrantRoundManager
-    //   - The round is active
-    for (uint256 i = 0; i < _rounds.length; i++) {
-      require(_rounds[i].isActive(), "GrantRoundManager: GrantRound is not active");
-      require(
-        donationToken == _rounds[i].donationToken(),
-        "GrantRoundManager: GrantRound's donation token does not match GrantRoundManager's donation token"
-      );
+  /**
+   * @dev Validates the the inputs to a donation call are valid, and reverts if any requirements are violated
+   * @param _donations Array of donations that will be executed
+   */
+  function _validateDonations(Donation[] calldata _donations) internal {
+    // TODO consider moving this to the section where we already loop through donations in case that saves a lot of
```

Comment:
are we saying 
- keep the function 
- let the argument be Donation _donation
- it gets invoked  as the first thing in _transferDonations 

It does seem a smart way to save gas but this would mean -> we'd do it after the swap happens which is I am not very comfortable about (but the argument could be made -> if you invoke it wrong -> it's not the contract's fault )

Reply:
Yea pretty much! The benefits of the current approach are that you save gas in the event of a failure (by reverting before the swap), but downside is more gas overall since you repeat a loop. I don't think the extra gas is significant, so I think we should we leave this as-is for readability, and come back to it later when we focus on optimizing gas usage